File: 2022/python/04_solution.py
def fully_contains(a_start: int, a_end: int, b_start: int, b_end: int) -> bool:
    if a_start >= b_start and a_end <= b_end:
        return True
    elif b_start >= a_start and b_end <= a_end:
        return True
    return False


def overlaps(a_start: int, a_end: int, b_start: int, b_end: int) -> bool:
    if a_end < b_start or b_end < a_start:
        return False
    return True


if __name__ == '__main__':
    with open('../input/04_input.txt', 'r', encoding='utf-8') as f:
        total1 = 0
        for line in f:
            a, b = line.strip().split(',')
            a_start, a_end = map(int, a.split('-'))
            b_start, b_end = map(int, b.split('-'))
            total1 += fully_contains(a_start, a_end, b_start, b_end)
    print(f'Part 1 - Fully contained assignment pairs: {total1}')

    with open('../input/04_input.txt', 'r', encoding='utf-8') as f:
        total2 = 0
        for line in f:
            a, b = line.strip().split(',')
            a_start, a_end = map(int, a.split('-'))
            b_start, b_end = map(int, b.split('-'))
            total2 += overlaps(a_start, a_end, b_start, b_end)
    print(f'Part 2 - Overlapping assignment pairs: {total2}')

File: 2022/ruby/04_solution.rb
# frozen_string_literal: true

def fully_contains(a_start, b_start, a_end, b_end)
  a_start >= b_start && a_end <= b_end || \
    b_start >= a_start && b_end <= a_end
end

def overlaps(a_start, b_start, a_end, b_end)
  a_end < b_start || b_end < a_start
end

if __FILE__ == $PROGRAM_NAME
  lines = IO.readlines('../input/04_input.txt', chomp: true)
  total1 = total2 = 0
  lines.each do |line|
    a, b = line.split(',')
    a_spl = a.split('-')
    b_spl = b.split('-')
    a_start = a_spl[0].to_i
    a_end = a_spl[-1].to_i
    b_start = b_spl[0].to_i
    b_end = b_spl[-1].to_i
    total1 += fully_contains(a_start, b_start, a_end, b_end) ? 1 : 0
    total2 += overlaps(a_start, b_start, a_end, b_end) ? 1 : 0
  end
  puts "Part 1 - Fully contained: #{total1}"
  puts "Part 2 - Overlapping: #{total2}"
end
